controllers/azurecluster_reconciler.go

```diff
@@ -114,7 +114,16 @@ func (s *azureClusterService) Delete(ctx context.Context) error {
 		return errors.Wrap(err, "failed to determine if the AzureCluster resource group is managed")
 	}
 	if managed {
-		// if the resource group is managed, we delete the entire resource group directly.
+		// If the resource group is managed, delete it.
+		// We need to explicitly delete vnet peerings, as it is not part of the resource group.
+		vnetPeeringsSvc, err := s.getService(vnetpeerings.ServiceName)
```

Comment:
Does an AzureCluster resource _always_ have an associated VNET peering service?

Reply:
It does not. I saw that the older version of this delete operation deleted resources like this [one by one](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2146/files#diff-5cf036f0f896f15c8f934e257474a75b1e3a4fd6f9e513a9fdfca962715564baL127), so I assumed that it would not return an error in the case of the resource not having a VNET peering service. I will test again to see if this is the intended behavior.

Reply:
If it is fine for a cluster not to have an associated VNET peering service, then I think we don't want to short-circuit the delete flow when we receive and error GET'ing a non-existent VNET peering service.

Reply:
I just tested this while deleting a cluster without a peering, and the error isn't reached if there are no vnet peerings on the resource. It doesn't short circuit the delete flow, so I think the error is only reached when the deletion fails. Would this be okay or should I handle the deletion in a different way?

Reply:
Was your test of a "managed" type, without a peering? I'm just reading through the code and it seems like for every such type, during deletion, we get the associated VNET peering service and return early with an error if the GET fails.

Reply:
Yes it was. I added debugging print statements within the code and the `getService` is not failing if there is no vnet peering. I could be understanding this incorrectly, but that's how it's working through my test.

It looks like the service list is being defined [here](https://github.com/willie-yao/cluster-api-provider-azure/blob/076783ee4a3010a414677efc88570e1201a87cf2/controllers/azurecluster_reconciler.go#L60) regardless of whether a particular service exists in the resource group or not, so `[getService](https://github.com/willie-yao/cluster-api-provider-azure/blob/076783ee4a3010a414677efc88570e1201a87cf2/controllers/azurecluster_reconciler.go#L143)` won't return an error.

Reply:
Delete operation is a no-op if no vnet peerings exist and is short-circuited here: https://github.com/willie-yao/cluster-api-provider-azure/blob/main/azure/services/vnetpeerings/vnetpeerings.go#L97

There is always a service defined, for every resource type, but each service short circuits reconcile/delete whenever there are no resources of that type for that AzureCluster, so this should be fine 👍